api/pre-1.44: Default `ReadOnlyNonRecursive` to true

vvoland opened the pull request:
- closes: https://github.com/moby/moby/issues/47355

Don't change the behavior for older clients and keep the same behavior.

**- How to verify it**
`TestContainerBindMountReadOnlyDefault`

**- Description for the changelog**
```markdown changelog
- To preserve backwards compatibility, read-only mounts are not recursive by default when using older clients (API version < v1.44).
```
**- A picture of a cute animal (not mandatory but encouraged)**

vvoland committed:
api/pre-1.44: Default `ReadOnlyNonRecursive` to true

Don't change the behavior for older clients and keep the same behavior.
Otherwise client can't opt-out (because `ReadOnlyNonRecursive` is
unsupported before 1.44).

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland commented on api/server/router/container/container_routes.go:
```diff
@@ -562,13 +562,20 @@ func (s *containerRouter) postContainersCreate(ctx context.Context, w http.Respo
 					return errdefs.InvalidParameter(errors.New("BindOptions.ReadOnlyForceRecursive needs API v1.44 or newer"))
+				} else {
+					bo = &mount.BindOptions{}
+					hostConfig.Mounts[idx].BindOptions = bo
```
Note: This is needed, because `m.BindOptions` would only overwrite the `BindOptions` in the `m` variable which is a local copy of the original mount.

vvoland commented:
Failure related, test needs adjustment:
```
=== Failed
=== FAIL: amd64.integration-cli TestDockerCLIRunSuite/TestRunMount (0.08s)
    docker_cli_run_test.go:4396: assertion failed: err is not nil: got error while creating a container with [--mount type=bind,src=/tmp/mount1181042325/mnt1,dst=/foo --mount type=bind,src=/tmp/mount1181042325/mnt2,dst=/bar] (mount-0-0)
    --- FAIL: TestDockerCLIRunSuite/TestRunMount (0.08s)
```

AkihiroSuda commented on api/server/router/container/container_routes.go:
```diff
@@ -562,13 +562,20 @@ func (s *containerRouter) postContainersCreate(ctx context.Context, w http.Respo
-				// Ignore ReadOnlyNonRecursive because it was added in API 1.44.
-				m.BindOptions.ReadOnlyNonRecursive = false
-				if m.BindOptions.ReadOnlyForceRecursive {
+		for idx, m := range hostConfig.Mounts {
```
hostConfig.Binds has to be checked too

vvoland replied:
Good catch, thanks!

AkihiroSuda reviewed (approved):
Thanks

vvoland commented:
Looks like some failures are related:
```
      Command:  /usr/local/cli-integration/docker run -d -v /vol1 -v /tmp/test-put-container-archive-err-symlink-in-volume-to-read-only-rootfs-1579682919:/vol2 -v /tmp/test-put-container-archive-err-symlink-in-volume-to-read-only-rootfs-1579682919/vol3:/vol3 -v /tmp/test-put-container-archive-err-symlink-in-volume-to-read-only-rootfs-1579682919/vol_ro:/vol_ro:ro --read-only busybox /bin/sh -c #(nop)
        ExitCode: 125
        Error:    exit status 125
        Stdout:   
        Stderr:   /usr/local/cli-integration/docker: Error response from daemon: invalid mount config for type "bind": bind source path does not exist: /tmp/test-put-container-archive-err-symlink-in-volume-to-read-only-rootfs-1579682919/vol_ro.
        See '/usr/local/cli-integration/docker run --help'.
```

thaJeztah commented on api/server/router/container/container_routes.go:
```diff
@@ -562,14 +563,45 @@ func (s *containerRouter) postContainersCreate(ctx context.Context, w http.Respo
+			if m.Type == mount.TypeBind && m.ReadOnly {
+				bo := m.BindOptions
+				if bo != nil {
+					if bo.ReadOnlyForceRecursive {
+						return errdefs.InvalidParameter(errors.New("BindOptions.ReadOnlyForceRecursive needs API v1.44 or newer"))
+					}
```
Perhaps something like this;

```go
for idx, m := range hostConfig.Mounts {
	if m.BindOptions != nil && m.BindOptions.ReadOnlyForceRecursive {
		// NOTE that technically this is a breaking change for older
		// API versions, and we should ignore the new field.
		// However, this option may be incorrectly set by a client with
		// the expectation that the failing to apply recursive read-only
		// is enforced, so we decided to produce an error instead,
		// instead of silently ignoring.
		return errdefs.InvalidParameter(errors.New("BindOptions.ReadOnlyForceRecursive needs API v1.44 or newer"))
	}
	if m.Type == mount.TypeBind && m.ReadOnly {
		if m.BindOptions == nil {
			hostConfig.Mounts[idx].BindOptions = &mount.BindOptions{}
		}
		// ReadOnlyNonRecursive was added in API 1.44.
		// Before that all read-only mounts were non-recursive.
		// Keep that behavior for clients on older APIs.
		hostConfig.Mounts[idx].BindOptions.ReadOnlyNonRecursive = true
	}
}
```

vvoland replied:
Hmm right, I guess it doesn't harm to do it for non-readonly case too (even though it doesn't really do anything without ReadOnly).

vvoland committed:
mounts/validate: Don't check source exists with CreateMountpoint

Don't error out when mount source doesn't exist and mounts has
`CreateMountpoint` option enabled.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland commented on api/server/router/container/container_routes.go:
```diff
@@ -562,14 +563,45 @@ func (s *containerRouter) postContainersCreate(ctx context.Context, w http.Respo
-				m.BindOptions.ReadOnlyNonRecursive = false
-				if m.BindOptions.ReadOnlyForceRecursive {
-					return errdefs.InvalidParameter(errors.New("BindOptions.ReadOnlyForceRecursive needs API v1.44 or newer"))
+		// Transform read-only binds into Mounts.
+		// This allows us to default ReadOnlyNonRecursive to true.
+		var newBinds []string
+		parser := mounts.NewParser()
```
We already error out in case there are duplicate mountpoints: https://github.com/moby/moby/blob/9d1541526c5beab51d2344b47624a160f199bbe6/daemon/volumes.go#L179

cpuguy83 replied:
We should not be doing this in the router at all.
I think it would be better to add an option to toggle which behavior to use and pass that down the stack.

vvoland replied:
Done

vvoland commented:
This looks green now, failures are unrelated and are caused by Docker Hub search API being down.

cpuguy83 commented on api/types/backend/backend.go:
```diff
@@ -18,6 +18,7 @@ type ContainerCreateConfig struct {
 	HostConfig       *container.HostConfig
 	NetworkingConfig *network.NetworkingConfig
 	Platform         *ocispec.Platform
+	TransformBinds   bool
```
```suggestion
	DefaultRecursiveReadOnly
```

cpuguy83 commented on daemon/create.go:
```diff
@@ -108,6 +111,11 @@ func (daemon *Daemon) containerCreate(ctx context.Context, daemonCfg *configStor
 		opts.params.HostConfig = &containertypes.HostConfig{}
 	}
+	if opts.params.TransformBinds {
+		if err := transformBinds(opts.params.HostConfig); err != nil {
```
I don't see why we are going to all this trouble here when we are already going to be parsing these mounts elsewhere, we can tell the it to set rro as a default based on the new param.

vvoland replied:
You're right, fixed!

cpuguy83 commented on integration/container/mounts_linux_test.go:
```diff
@@ -426,6 +426,69 @@ func TestContainerCopyLeaksMounts(t *testing.T) {
 	assert.Equal(t, mountsBefore, mountsAfter)
 }
 
+func TestContainerBindMountReadOnlyDefault(t *testing.T) {
```
I think we need a to check if rro is supported on the kernel for the test to function properly.

Alternatively, we could just check the registered mountpoints in the API to see if the option is set as expected.

cpuguy83 replied:
Also sorry, one more check, test should be skipped if api version is < 1.44.

vvoland replied:
Why? It also tests that the behavior doesn't change when <1.44 API is requested.

cpuguy83 replied:
So yes technically the 1.43 test should pass on an older daemon.
But the 1.44 test would fail since it is not a supported version.

vvoland replied:
1.44 test will be skipped: https://github.com/moby/moby/pull/47391/files#diff-0ac47068153ec59b77b2044bd459fb6395f028e6e112dc03645f7ba72f324bf2R456

But yeah the `version: ""` would fail, updated it to be also skipped.

Also, added the test for `api.MinSupportedAPIVersion` client version.

thaJeztah reviewed (approved):
LGTM, thanks!
